### ISSUE TITLE
microVM: stdio multiplexing for many concurrent sandboxes

### ISSUE THREAD

**Issue by peterp** [redwoodjs/machinen#51]
## Problem

Right now one VM owns the host terminal — raw mode, keystrokes straight through the PL011 stdin thread to Node. Once you have N sandboxes running, you can't share one terminal across them. There's no way to address "talk to sandbox #3" without tearing down the current stdio model.

## Goal

Run many sandboxes at once with clean, per-sandbox stdio, addressable from the host.

## Scope

- Per-sandbox pty pair on the host, with one end attached to the VM's stdio channel.
- A supervisor process that owns the real terminal and demultiplexes: switch between sandboxes, scrollback, logs to disk, attach/detach.
- Probably virtio-vsock under the hood instead of (or alongside) PL011 — scales better, doesn't need per-VM tty emulation, cleaner protocol for a host-side agent.
- CLI/library surface: `machinen.list()`, `machinen.attach(id)`, `machinen.detach()`.

## Depends on

The spawn API (#49) — without it there's no "many sandboxes" to mux. Also unblocks the nice-UX version of CC-in-a-sandbox (#48).

## Done when

With three sandboxes booted via `machinen.spawn()`, you can attach to any one interactively, send input, see only that sandbox's output, detach, attach to another — all without restarting any of them.

## Refs

- #42, #48 (Claude Code), #49 (spawn API).

**Comment by peterp** [comment]
No code yet — design doc in \`.docs/learnings/microvm/stdio-mux-plan.md\`.

Dependencies that have since landed in the meantime:
- #49 + #50 M2: \`@machinen/runtime\` now has a \`spawn\` API with per-VM stdin/stdout/stderr handles. That's the primitive the supervisor will multiplex.
- #44 transport: the virtio-MMIO work from #46 / #47 makes adding virtio-vsock a small increment when we want it.

Not starting M1 until there's a real workflow that feels the single-terminal pain (two concurrent CC sandboxes, for instance). Right now everything runs one-at-a-time via \`test-fixtures/try.sh\` modes and that's enough.

**Comment by peterp** [comment]
### M1 done

Committed in 76570be. Registry + supervisor shipped; PTY niceties deliberately deferred.

### What landed

- `@machinen/runtime` now exports `Sandboxes` and `Supervisor`.
- **`Sandboxes`** — registry with `add(id, vm)` / `remove` / `list` / `get` / `send(id, bytes)` / `onOutput(id, fn)`. Auto-removes entries when the VMM exits. Bounded per-sandbox scrollback ring (8 KiB default) so a late attach replays recent context without unbounded memory growth.
- **`Supervisor`** — owns a `Readable` + `Writable` pair (default `process.stdin`/`process.stdout`). Detached mode parses slash-commands (`/ls`, `/attach <id>`, `/help`, `/quit`); attached mode pipes bytes verbatim. `Ctrl-] Ctrl-]` (two 0x1D bytes) detaches.
- Tiny manual driver at `packages/runtime/src/bin/supervise.ts` — spawns N `/bin/cat` children and hands the TTY to the supervisor, useful for eyeballing the UX.

### Tests

`pnpm vitest run packages/runtime` = **17/17**. Multiplex adds 9:

- Registry CRUD + duplicate rejection
- `onOutput` fires for new bytes; scrollback preserves history
- Scrollback bounded correctly
- Entry auto-removed on VM exit
- Supervisor forwards typed bytes to the attached sandbox
- Routing targets the currently-attached sandbox only
- `Ctrl-] Ctrl-]` detaches
- `/ls` lists ids

### What's still ahead

- **PTY layer.** Swap the per-VMM pipes for pty pairs so Node/CC see a real terminal under attach (colors, `TIOCSWINSZ` on resize). Native dep (`node-pty`) or a tiny Zig helper that does `openpty`/`forkpty` and exposes the master over a unix socket.
- **vsock control plane (#44).** Once the transport lands, the Supervisor can run introspection commands (list guest processes, tail a log, ship a file in) without stealing the stdio stream.
- **CLI wrapper.** `@machinen/cli` command that wraps `Supervisor` + a one-line sandbox spawn.

**Comment by peterp** [comment]
M1.5 landed: commit 1577d89 on feat/vmm-zig-mvp.

**What shipped**:
- **`spawnPty(opts) -> PtyVmHandle`** in `@machinen/runtime` — same shape as `spawn()` plus `resize(cols, rows)`. Under the hood it's node-pty (prebuilt arm64 macOS + linux), wrapped with a one-shot `ensureSpawnHelper()` that chmods the helper binary pnpm sometimes leaves without the exec bit.
- **Supervisor terminal integration**:
  - Raw-mode stdin on attach, restored on detach (default-on when input is a TTY).
  - SIGWINCH → `entry.vm.resize(cols, rows)` forwarding, so the attached sandbox's pty reshapes live when you resize the host window. Pushes once on attach so the initial size is right.
- **`supervise.ts` demo** now spawns `$SHELL -i` per sandbox under a pty — makes it easy to feel the difference vs the old pipe path.

**Tests**: 21/21 pass (17 existing + 4 new):
- `tty` inside the child emits a pty device path.
- `stty size` reflects initial + resized shape.
- `Sandboxes.add` works with PtyVmHandle; scrollback captures pty output.
- Supervisor calls `resize()` on SIGWINCH emission.

Hit one subtle bug worth flagging in the learnings: a naive internal `.on('data', …)` listener on the PassThrough drains it before the real consumer attaches. Fixed by collecting inside node-pty's `onData` callback (before any stream listener) and leaving the PassThrough paused until Sandboxes / Supervisor / caller subscribes.

**Still deferred** (documented in [`.docs/learnings/microvm/multiplex-pty.md`](https://github.com/redwoodjs/machinen/blob/feat/vmm-zig-mvp/.docs/learnings/microvm/multiplex-pty.md)):
- **Guest-side TIOCSWINSZ.** Host pty resize doesn't propagate into the guest (which sees ttyAMA0, not a pty). Needs a vsock-based agent path — low-hanging fruit now that vsock M2+ is in (#44).
- **vsock control plane + CLI wrapper** — still out of scope for this milestone.

Proposing M1.5 closes this one out unless the vsock-forwarded resize is worth tracking here.

**Comment by peterp** [comment]
Guest-side TIOCSWINSZ landed as the natural follow-up: commit 36efb03.

**What's new**:
- **`packages/microvm/test-fixtures/winsize-agent.py`** — AF_VSOCK listener on port 1974 inside the guest; ioctls TIOCSWINSZ on `/dev/console`, `/dev/ttyAMA0`, `/dev/tty0`, and every tty-owning fd under `/proc`. That /proc sweep is what makes long-lived shells actually see the new size (they captured it at boot via TIOCGWINSZ and wouldn't re-query).
- **`VsockWinsize` in `@machinen/runtime`** — 2-method UDS client: `.send(cols, rows)` + `.close()`. `VsockWinsize.connect(path, { timeoutMs })` retries until the bridge publishes the socket so app code can call it right after `spawn()`. Dedupes same-size sends.

**Smoke** (`smoke.sh winsize`): host sends 80x24 then 132x50 → agent acks both on the vsock channel → guest log shows ioctl success on /dev/console for both sizes. All 3 assertions pass.

**End-to-end of what actually runs**:
```
host Python  -UDS->  vsock bridge  -virtio->  guest kernel
  -AF_VSOCK->  agent.py  -ioctl->  /dev/console (+ttyAMA0, tty0, PIDs)
```

**Wiring into the Supervisor** — `.docs/learnings/microvm/multiplex-pty.md` now documents the 4-line pattern: boot with `MACHINEN_VSOCK=in:1974:/tmp/winsize.sock`, `await VsockWinsize.connect(...)`, override the handle's `.resize` to fan out. Supervisor's `forwardResize` path does the rest.

With this, the terminal story for #51 is end-to-end: colors, raw mode, host SIGWINCH, guest ttys all tracking the real window size. Nothing still open that I'd block on.

**Comment by peterp** [comment]
Multiplex end-to-end: registry + supervisor (76570be), PTY + SIGWINCH forwarding (1577d89), guest-side TIOCSWINSZ via vsock agent (36efb03). `supervise.ts` demo attaches/detaches across sandboxes. vsock-based control plane beyond resize is a separate concern.
